File: 7_graph/bfs求无权图的最短路径/getPath.go
package main

// 求from到to的一条路径.
func GetPath(graph [][]int32, from, to int32) []int32 {
	n := int32(len(graph))
	visited := make([]bool, n)
	visited[from] = true
	stack := []int32{from}
	pre := make([]int32, n)
	for len(stack) > 0 {
		cur := stack[len(stack)-1]
		if cur == to {
			break
		}
		stack = stack[:len(stack)-1]
		for _, next := range graph[cur] {
			if !visited[next] {
				visited[next] = true
				pre[next] = cur
				stack = append(stack, next)
			}
		}
	}
	if !visited[to] {
		return nil
	}
	path := []int32{to}
	for v := to; v != from; v = pre[v] {
		path = append(path, pre[v])
	}
	for i := 0; i < len(path)/2; i++ {
		path[i], path[len(path)-1-i] = path[len(path)-1-i], path[i]
	}
	return path
}

File: 7_graph/bfs求无权图的最短路径/getPath.py
from typing import List


def getPath(graph: List[List[int]], start: int, target: int) -> List[int]:
    n = len(graph)
    visited = [False] * n
    pre = [-1] * n
    stack = [start]
    visited[start] = True
    while stack:
        cur = stack.pop()
        if cur == target:
            break
        for next_ in graph[cur]:
            if not visited[next_]:
                visited[next_] = True
                pre[next_] = cur
                stack.append(next_)
    if not visited[target]:
        return []
    res = []
    while target != -1:
        res.append(target)
        target = pre[target]
    return res[::-1]


if __name__ == "__main__":
    print(getPath([[1, 2], [3], [3], []], 0, 3))  # [0, 2, 3]
    print(getPath([[1, 2], [3], [3], []], 0, 1))  # [0, 1]
    print(getPath([[1, 2], [3], [3], []], 0, 0))  # [0]
    print(getPath([[1, 2], [3], [3], []], 0, 2))  # [0, 2]
